### ISSUE TITLE
Add test to avoid unwanted dependencies among backends

### ISSUE THREAD

**Issue by datapythonista** [ibis-project/ibis#2494]
Ibis had some interdependencies among backends that have been causing some issues, besides making the codebase unnecessarily complicated.

We are finishing a major refactoring to make the dependencies more reasonable. Mainly, to avoid unrelated backends to depend among them (bigquery or spark depending on impala), and creating some base backends with the code shared among them.

I think it would be great if we can have a test that makes sure there are not unwanted dependencies among backends.

I think we can probably have a list of the backends with the known dependencies, load each backend, and check if any unknown backend has been loaded. As a general idea (code not expected to work):
```python
BACKEND_DEPS = {'sqlite': ['base_sqlalchemy'],
                'bigquery': ['base_sql'],
                ...}

for backend, known_deps in BACKEND_DEPS:
    importlib.load(f'ibis.backends.{backend}')
    assert not any(mod in sys.modules if mod.startswith('ibis.backends') and mod not in known_deps)
    importlib.unload(...)
```

**Comment by matthewmturner** [comment]
this would be good to get in the next release along with the refactoring, right?

i was thinking the following:

1. new `tests` folder in `ibis/backends`
2. move `ibis/backends/conftest.py` to `ibis/backends/tests`
3. create `test_dependencies.py` in `ibis/backends/tests` with implementation of the logic you provided

I can try to do this in parallel to operation registry move to confirm expected results by testing before the final batch move and then after.

what do you think?

**Comment by datapythonista** [comment]
Thanks @matthewmturner that sounds like good ideas. But I don't think we can do number 2. If you move the `conftest.py` file, I think it won't be applied to the backend tests then, which is it's goal.

This is a nice to have, but surely not a blocker, and lower priority than the refactoring itself, but if you feel like working on it, it'll be useful to know when we're done with the refactoring. :)

**Comment by matthewmturner** [comment]
ive been looking into this.  from what ive seen there isnt a way to unload a module(i.e. what `importlib.unload` from your example was representing).  The closest i saw was `del sys.modules["module"]`.  i suppose we could loop through the known deps and delete them that way.  or do you think it makes sense to just add a `test_dependencies.py` to each backend and do the appropriate test?

**Comment by datapythonista** [comment]
If `del sys.modules...` works I'd go for that. Adding tests for each backends is very repetitive, and we'll end up missing backends in the future.

**Comment by matthewmturner** [comment]
Something i found interesting.  When importing `ibis.backends.sqlite` i end up with `ibis.backends.impala`,  `ibis.backends.spark`,  and `ibis.backends.pyspark` modules in `sys.modules`.  Havent gotten to the bottom of this yet but let me know if you have any ideas.

Also, i assume pandas is fine is a known backend for all backends?  and what about the file based backends (base_file, csv, parquet, hdf5)

```
Python 3.8.5 | packaged by conda-forge | (default, Sep 24 2020, 16:55:52)
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; import ibis.backends.sqlite
>>> [mod for mod in sys.modules.keys() if mod.startswith("ibis.backends")]
['ibis.backends', 'ibis.backends.base_sqlalchemy', 'ibis.backends.base_sqlalchemy.transforms', 'ibis.backends.base_sqlalchemy.compiler', 'ibis.backends.pandas.aggcontext', 'ibis.backends.pandas.dispatcher', 'ibis.backends.pandas.trace', 'ibis.backends.pandas.dispatch', 'ibis.backends.pandas.core', 'ibis.backends.pandas.client', 'ibis.backends.pandas.execution.arrays', 'ibis.backends.pandas.execution.decimal', 'ibis.backends.pandas.execution.constants', 'ibis.backends.pandas.execution.util', 'ibis.backends.pandas.execution.generic', 'ibis.backends.pandas.execution.join', 'ibis.backends.pandas.execution.maps', 'ibis.backends.pandas.execution.selection', 'ibis.backends.pandas.execution.strings', 'ibis.backends.pandas.execution.structs', 'ibis.backends.pandas.execution.temporal', 'ibis.backends.pandas.execution.timecontext', 'ibis.backends.pandas.execution.window', 'ibis.backends.pandas.execution', 'ibis.backends.pandas.udf', 'ibis.backends.pandas', 'ibis.backends.base_file', 'ibis.backends.csv', 'ibis.backends.parquet', 'ibis.backends.hdf5', 'ibis.backends.base_sql.identifiers', 'ibis.backends.base_sql', 'ibis.backends.base_sql.ddl', 'ibis.backends.impala.ddl', 'ibis.backends.base_sql.compiler', 'ibis.backends.impala.compiler', 'ibis.backends.impala.udf', 'ibis.backends.base_sqlalchemy.alchemy', 'ibis.backends.sqlite.compiler', 'ibis.backends.sqlite.client', 'ibis.backends.sqlite', 'ibis.backends.spark.compiler', 'ibis.backends.spark.ddl', 'ibis.backends.spark.datatypes', 'ibis.backends.spark.client', 'ibis.backends.spark.udf', 'ibis.backends.spark', 'ibis.backends.pyspark.operations', 'ibis.backends.pyspark.timecontext', 'ibis.backends.pyspark.compiler', 'ibis.backends.pyspark.client', 'ibis.backends.pyspark']
```

**Comment by datapythonista** [comment]
I think we probably need to merge #2379 before this can work. At the moment, all backends that don't fail to load (in theory because of missing dependencies) are loaded in `ibis/__init__.py`. So I think this is what you're seeing here.

**Comment by matthewmturner** [comment]
> 
> 
> I think we probably need to merge #2379 before this can work. At the moment, all backends that don't fail to load (in theory because of missing dependencies) are loaded in `ibis/__init__.py`. So I think this is what you're seeing here.

ah got it. makes sense. thanks.
